File: AnimForgeLiveLink/UnrealGymProject/Source/Programs/AnimForgeMayaWarpViz/Scripts/warpviz_protocol.py
"""AnimForge WarpViz wire protocol - Python mirror.

This module is a 1:1 mirror of Shared/AnimForgeWarpVizShared/WarpVizProtocol.h
(.cpp). It is used by:

  * the automated test suite (AnimForgeLiveLink/Tests/Python),
  * mock_unreal_server.py - a stand-in for the Unreal gym so the Maya side of
    the pipeline can be exercised without an engine session,
  * optional Maya-side Python tooling that wants to talk to the gym directly.

Frame layout (little-endian):
    [4 bytes magic b"AFWV"] [4 bytes payload length] [payload: UTF-8 JSON]

Envelope:
    {"type": "<MessageType>", "protocolVersion": 1, "payload": {...}}

Keep this file dependency-free (stdlib only) so it runs inside mayapy,
a plain CPython, and CI without any pip installs.
"""

from __future__ import annotations

import json
import struct
import uuid
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

PROTOCOL_VERSION = 1
DEFAULT_PORT = 46464
FRAME_MAGIC = b"AFWV"
MAX_PAYLOAD_BYTES = 256 * 1024 * 1024

WARP_METHODS = ("SkewWarp", "SimpleWarp", "Scale")

MSG_HANDSHAKE = "Handshake"
MSG_HANDSHAKE_ACK = "HandshakeAck"
MSG_EVALUATE_REQUEST = "EvaluateRequest"
MSG_EVALUATE_PROGRESS = "EvaluateProgress"
MSG_EVALUATE_RESULT = "EvaluateResult"
MSG_ERROR = "Error"


class ProtocolError(ValueError):
    """Raised when a payload fails validation."""


# ---------------------------------------------------------------------------
# Framing
# ---------------------------------------------------------------------------

def encode_frame(payload: bytes) -> bytes:
    """Wraps a UTF-8 JSON payload in [magic][length][payload]."""
    return FRAME_MAGIC + struct.pack("<I", len(payload)) + payload


class FrameDecoder:
    """Incremental frame decoder mirroring the C++ FrameDecoder.

    Feed raw socket bytes in arbitrary fragmentation; iterate next() until it
    returns None. Resynchronizes on the magic when the stream is corrupted.
    """

    def __init__(self) -> None:
        self._buffer = bytearray()
        self.dropped_bytes = 0

    def feed(self, data: bytes) -> None:
        self._buffer.extend(data)

    def _resync(self) -> None:
        drop = 0
        n = len(self._buffer)
        while drop < n:
            compare = min(4, n - drop)
            if self._buffer[drop:drop + compare] == FRAME_MAGIC[:compare]:
                break
            drop += 1
        if drop:
            del self._buffer[:drop]
            self.dropped_bytes += drop

    def next(self) -> Optional[bytes]:
        """Pops the next complete payload, or None if no full frame is buffered."""
        while True:
            self._resync()
            if len(self._buffer) < 8:
                return None
            (length,) = struct.unpack_from("<I", self._buffer, 4)
            if length > MAX_PAYLOAD_BYTES:
                # Corrupt length: skip this magic and resync further on.
                del self._buffer[:4]
                self.dropped_bytes += 4
                continue
            if len(self._buffer) < 8 + length:
                return None
            payload = bytes(self._buffer[8:8 + length])
            del self._buffer[:8 + length]
            return payload


# ---------------------------------------------------------------------------
# Message dataclasses
# ---------------------------------------------------------------------------

Vec3 = Tuple[float, float, float]
Quat = Tuple[float, float, float, float]  # (x, y, z, w)

QUAT_IDENTITY: Quat = (0.0, 0.0, 0.0, 1.0)


@dataclass
class TimeRange:
    start_frame: float = 0.0
    end_frame: float = 0.0
    fps: float = 30.0

    def duration_seconds(self) -> float:
        return (self.end_frame - self.start_frame) / self.fps if self.fps > 0 else 0.0

    def to_json(self) -> Dict:
        return {"startFrame": self.start_frame, "endFrame": self.end_frame, "fps": self.fps}

    @classmethod
    def from_json(cls, data: Dict) -> "TimeRange":
        return cls(
            start_frame=float(data.get("startFrame", 0.0)),
            end_frame=float(data.get("endFrame", 0.0)),
            fps=float(data.get("fps", 30.0)),
        )


@dataclass
class TrajectorySample:
    time_seconds: float
    translation: Vec3
    rotation: Quat = QUAT_IDENTITY

    def to_json(self) -> Dict:
        return {"t": self.time_seconds, "pos": list(self.translation), "rot": list(self.rotation)}

    @classmethod
    def from_json(cls, data: Dict) -> "TrajectorySample":
        pos = data.get("pos")
        rot = data.get("rot")
        if not isinstance(pos, list) or len(pos) != 3:
            raise ProtocolError("trajectory sample 'pos' must be a 3-array")
        if not isinstance(rot, list) or len(rot) != 4:
            raise ProtocolError("trajectory sample 'rot' must be a 4-array")
        return cls(float(data.get("t", 0.0)), tuple(map(float, pos)), tuple(map(float, rot)))


@dataclass
class WarpTarget:
    name: str = ""
    translation: Vec3 = (0.0, 0.0, 0.0)
    rotation: Quat = QUAT_IDENTITY

    def to_json(self) -> Dict:
        return {"name": self.name, "pos": list(self.translation), "rot": list(self.rotation)}

    @classmethod
    def from_json(cls, data: Dict) -> "WarpTarget":
        pos = data.get("pos")
        rot = data.get("rot")
        if not isinstance(pos, list) or len(pos) != 3 or not isinstance(rot, list) or len(rot) != 4:
            raise ProtocolError("warp target transform malformed")
        return cls(str(data.get("name", "")), tuple(map(float, pos)), tuple(map(float, rot)))


@dataclass
class GhostPose:
    time_seconds: float
    joints: List[Dict] = field(default_factory=list)  # {"name", "pos", "rot"}

    def to_json(self) -> Dict:
        return {"t": self.time_seconds, "joints": self.joints}

    @classmethod
    def from_json(cls, data: Dict) -> "GhostPose":
        return cls(float(data.get("t", 0.0)), list(data.get("joints", [])))


@dataclass
class EvaluateRequest:
    request_id: str
    character_id: str
    clip_name: str
    range: TimeRange
    warp_window: TimeRange
    method: str = "SkewWarp"
    target: WarpTarget = field(default_factory=WarpTarget)
    warp_rotation: bool = True
    warp_translation: bool = True
    ghost_interval_frames: float = 5.0
    maya_root_samples: List[TrajectorySample] = field(default_factory=list)

    @staticmethod
    def new_request_id() -> str:
        return uuid.uuid4().hex


@dataclass
class EvaluateResult:
    request_id: str
    success: bool = False
    error_message: str = ""
    warped_trajectory: List[TrajectorySample] = field(default_factory=list)
    original_trajectory: List[TrajectorySample] = field(default_factory=list)
    ghost_poses: List[GhostPose] = field(default_factory=list)
    warnings: List[str] = field(default_factory=list)
    evaluation_ms: float = 0.0


# ---------------------------------------------------------------------------
# Envelope encode / decode
# ---------------------------------------------------------------------------

def _envelope(msg_type: str, payload: Dict) -> bytes:
    doc = {"type": msg_type, "protocolVersion": PROTOCOL_VERSION, "payload": payload}
    return json.dumps(doc, separators=(",", ":")).encode("utf-8")


def peek_envelope(payload: bytes) -> Tuple[str, int]:
    """Returns (type, protocolVersion) without validating the payload body."""
    doc = json.loads(payload.decode("utf-8"))
    return str(doc.get("type", "Unknown")), int(doc.get("protocolVersion", 0))


def _open_envelope(payload: bytes, expected_type: str) -> Dict:
    try:
        doc = json.loads(payload.decode("utf-8"))
    except (ValueError, UnicodeDecodeError) as exc:
        raise ProtocolError("payload is not valid JSON: %s" % exc)
    version = int(doc.get("protocolVersion", 0))
    if version != PROTOCOL_VERSION:
        raise ProtocolError(
            "protocol version mismatch: got %d, expected %d" % (version, PROTOCOL_VERSION))
    if doc.get("type") != expected_type:
        raise ProtocolError(
            "unexpected message type: got %r, expected %r" % (doc.get("type"), expected_type))
    body = doc.get("payload")
    if not isinstance(body, dict):
        raise ProtocolError("missing 'payload' object")
    return body


# --- Handshake --------------------------------------------------------------

def encode_handshake(client_name: str, character_id: str) -> bytes:
    return _envelope(MSG_HANDSHAKE, {"clientName": client_name, "characterId": character_id})


def decode_handshake(payload: bytes) -> Dict:
    body = _open_envelope(payload, MSG_HANDSHAKE)
    return {"client_name": body.get("clientName", ""), "character_id": body.get("characterId", "")}


def encode_handshake_ack(server_name: str, known_clips: List[str]) -> bytes:
    return _envelope(MSG_HANDSHAKE_ACK, {"serverName": server_name, "knownClips": list(known_clips)})


def decode_handshake_ack(payload: bytes) -> Dict:
    body = _open_envelope(payload, MSG_HANDSHAKE_ACK)
    return {
        "server_name": body.get("serverName", ""),
        "known_clips": [c for c in body.get("knownClips", []) if isinstance(c, str)],
    }


# --- EvaluateRequest ---------------------------------------------------------

def encode_evaluate_request(req: EvaluateRequest) -> bytes:
    return _envelope(MSG_EVALUATE_REQUEST, {
        "requestId": req.request_id,
        "characterId": req.character_id,
        "clipName": req.clip_name,
        "range": req.range.to_json(),
        "warpWindow": req.warp_window.to_json(),
        "method": req.method,
        "target": req.target.to_json(),
        "warpRotation": req.warp_rotation,
        "warpTranslation": req.warp_translation,
        "ghostIntervalFrames": req.ghost_interval_frames,
        "mayaRootSamples": [s.to_json() for s in req.maya_root_samples],
    })


def decode_evaluate_request(payload: bytes) -> EvaluateRequest:
    body = _open_envelope(payload, MSG_EVALUATE_REQUEST)

    request_id = str(body.get("requestId", ""))
    if not request_id:
        raise ProtocolError("EvaluateRequest missing 'requestId'")
    method = str(body.get("method", ""))
    if method not in WARP_METHODS:
        raise ProtocolError("EvaluateRequest has unknown warp method %r" % method)
    if not isinstance(body.get("range"), dict):
        raise ProtocolError("EvaluateRequest missing 'range'")
    if not isinstance(body.get("target"), dict):
        raise ProtocolError("EvaluateRequest missing 'target'")

    rng = TimeRange.from_json(body["range"])
    window = TimeRange.from_json(body["warpWindow"]) if isinstance(body.get("warpWindow"), dict) else rng

    return EvaluateRequest(
        request_id=request_id,
        character_id=str(body.get("characterId", "")),
        clip_name=str(body.get("clipName", "")),
        range=rng,
        warp_window=window,
        method=method,
        target=WarpTarget.from_json(body["target"]),
        warp_rotation=bool(body.get("warpRotation", True)),
        warp_translation=bool(body.get("warpTranslation", True)),
        ghost_interval_frames=float(body.get("ghostIntervalFrames", 5.0)),
        maya_root_samples=[TrajectorySample.from_json(s) for s in body.get("mayaRootSamples", [])],
    )


# --- EvaluateProgress --------------------------------------------------------

def encode_evaluate_progress(request_id: str, progress: float, stage: str) -> bytes:
    return _envelope(MSG_EVALUATE_PROGRESS,
                     {"requestId": request_id, "progress": progress, "stage": stage})


def decode_evaluate_progress(payload: bytes) -> Dict:
    body = _open_envelope(payload, MSG_EVALUATE_PROGRESS)
    return {
        "request_id": body.get("requestId", ""),
        "progress": float(body.get("progress", 0.0)),
        "stage": body.get("stage", ""),
    }


# --- EvaluateResult ----------------------------------------------------------

def encode_evaluate_result(res: EvaluateResult) -> bytes:
    return _envelope(MSG_EVALUATE_RESULT, {
        "requestId": res.request_id,
        "success": res.success,
        "errorMessage": res.error_message,
        "warpedTrajectory": [s.to_json() for s in res.warped_trajectory],
        "originalTrajectory": [s.to_json() for s in res.original_trajectory],
        "ghostPoses": [g.to_json() for g in res.ghost_poses],
        "warnings": list(res.warnings),
        "evaluationMs": res.evaluation_ms,
    })


def decode_evaluate_result(payload: bytes) -> EvaluateResult:
    body = _open_envelope(payload, MSG_EVALUATE_RESULT)
    return EvaluateResult(
        request_id=str(body.get("requestId", "")),
        success=bool(body.get("success", False)),
        error_message=str(body.get("errorMessage", "")),
        warped_trajectory=[TrajectorySample.from_json(s) for s in body.get("warpedTrajectory", [])],
        original_trajectory=[TrajectorySample.from_json(s) for s in body.get("originalTrajectory", [])],
        ghost_poses=[GhostPose.from_json(g) for g in body.get("ghostPoses", [])],
        warnings=[w for w in body.get("warnings", []) if isinstance(w, str)],
        evaluation_ms=float(body.get("evaluationMs", 0.0)),
    )


# --- Error -------------------------------------------------------------------

def encode_error(message: str, request_id: str = "") -> bytes:
    return _envelope(MSG_ERROR, {"requestId": request_id, "message": message})


def decode_error(payload: bytes) -> Dict:
    body = _open_envelope(payload, MSG_ERROR)
    return {"request_id": body.get("requestId", ""), "message": body.get("message", "")}

File: AnimForgeLiveLink/UnrealGymProject/Shared/AnimForgeWarpVizShared/WarpVizProtocol.cpp
// AnimForgeWarpVizShared - WarpVizProtocol.cpp

#include "WarpVizProtocol.h"

#include <cstring>

namespace AnimForge
{
namespace WarpViz
{

// ---------------------------------------------------------------------------
// Enum <-> string
// ---------------------------------------------------------------------------

const char* WarpMethodToString(WarpMethod Method)
{
    switch (Method)
    {
    case WarpMethod::SkewWarp:   return "SkewWarp";
    case WarpMethod::SimpleWarp: return "SimpleWarp";
    case WarpMethod::Scale:      return "Scale";
    default:                     return "Unknown";
    }
}

WarpMethod WarpMethodFromString(const std::string& Name)
{
    if (Name == "SkewWarp")   return WarpMethod::SkewWarp;
    if (Name == "SimpleWarp") return WarpMethod::SimpleWarp;
    if (Name == "Scale")      return WarpMethod::Scale;
    return WarpMethod::Unknown;
}

const char* MessageTypeToString(MessageType Type)
{
    switch (Type)
    {
    case MessageType::Handshake:        return "Handshake";
    case MessageType::HandshakeAck:     return "HandshakeAck";
    case MessageType::EvaluateRequest:  return "EvaluateRequest";
    case MessageType::EvaluateProgress: return "EvaluateProgress";
    case MessageType::EvaluateResult:   return "EvaluateResult";
    case MessageType::Error:            return "Error";
    default:                            return "Unknown";
    }
}

MessageType MessageTypeFromString(const std::string& Name)
{
    if (Name == "Handshake")        return MessageType::Handshake;
    if (Name == "HandshakeAck")     return MessageType::HandshakeAck;
    if (Name == "EvaluateRequest")  return MessageType::EvaluateRequest;
    if (Name == "EvaluateProgress") return MessageType::EvaluateProgress;
    if (Name == "EvaluateResult")   return MessageType::EvaluateResult;
    if (Name == "Error")            return MessageType::Error;
    return MessageType::Unknown;
}

// ---------------------------------------------------------------------------
// Framing
// ---------------------------------------------------------------------------

std::vector<uint8_t> EncodeFrame(const std::string& Payload)
{
    std::vector<uint8_t> Frame;
    Frame.reserve(8 + Payload.size());
    Frame.insert(Frame.end(), kFrameMagic, kFrameMagic + 4);

    const uint32_t Length = static_cast<uint32_t>(Payload.size());
    Frame.push_back(static_cast<uint8_t>(Length & 0xFF));
    Frame.push_back(static_cast<uint8_t>((Length >> 8) & 0xFF));
    Frame.push_back(static_cast<uint8_t>((Length >> 16) & 0xFF));
    Frame.push_back(static_cast<uint8_t>((Length >> 24) & 0xFF));

    Frame.insert(Frame.end(), Payload.begin(), Payload.end());
    return Frame;
}

void FrameDecoder::Feed(const uint8_t* Data, size_t Length)
{
    Buffer.insert(Buffer.end(), Data, Data + Length);
}

void FrameDecoder::Resync()
{
    // Drop bytes until the buffer starts with the magic (or could - a partial
    // magic at the tail is kept so a split magic still decodes).
    size_t Drop = 0;
    while (Drop < Buffer.size())
    {
        const size_t Remaining = Buffer.size() - Drop;
        const size_t Compare = Remaining < 4 ? Remaining : 4;
        if (std::memcmp(Buffer.data() + Drop, kFrameMagic, Compare) == 0)
        {
            break;
        }
        ++Drop;
    }
    if (Drop > 0)
    {
        Buffer.erase(Buffer.begin(), Buffer.begin() + Drop);
        TotalDroppedBytes += Drop;
    }
}

bool FrameDecoder::Next(std::string& OutPayload)
{
    Resync();
    if (Buffer.size() < 8)
    {
        return false;
    }

    const uint32_t Length =
        static_cast<uint32_t>(Buffer[4])
        | (static_cast<uint32_t>(Buffer[5]) << 8)
        | (static_cast<uint32_t>(Buffer[6]) << 16)
        | (static_cast<uint32_t>(Buffer[7]) << 24);

    if (Length > kMaxPayloadBytes)
    {
        // Corrupt length: skip this magic and try to resync further on.
        Buffer.erase(Buffer.begin(), Buffer.begin() + 4);
        TotalDroppedBytes += 4;
        return Next(OutPayload);
    }

    if (Buffer.size() < 8 + static_cast<size_t>(Length))
    {
        return false; // frame incomplete, wait for more bytes
    }

    OutPayload.assign(reinterpret_cast<const char*>(Buffer.data() + 8), Length);
    Buffer.erase(Buffer.begin(), Buffer.begin() + 8 + Length);
    return true;
}

// ---------------------------------------------------------------------------
// JSON helpers
// ---------------------------------------------------------------------------

JsonValue Vec3ToJson(const Vec3& V)
{
    JsonValue Arr = JsonValue::MakeArray();
    Arr.Add(JsonValue(V.X));
    Arr.Add(JsonValue(V.Y));
    Arr.Add(JsonValue(V.Z));
    return Arr;
}

JsonValue QuatToJson(const Quat& Q)
{
    JsonValue Arr = JsonValue::MakeArray();
    Arr.Add(JsonValue(Q.X));
    Arr.Add(JsonValue(Q.Y));
    Arr.Add(JsonValue(Q.Z));
    Arr.Add(JsonValue(Q.W));
    return Arr;
}

bool Vec3FromJson(const JsonValue& J, Vec3& Out)
{
    if (!J.IsArray() || J.Items().size() != 3)
    {
        return false;
    }
    Out.X = J.Items()[0].AsNumber();
    Out.Y = J.Items()[1].AsNumber();
    Out.Z = J.Items()[2].AsNumber();
    return true;
}

bool QuatFromJson(const JsonValue& J, Quat& Out)
{
    if (!J.IsArray() || J.Items().size() != 4)
    {
        return false;
    }
    Out.X = J.Items()[0].AsNumber();
    Out.Y = J.Items()[1].AsNumber();
    Out.Z = J.Items()[2].AsNumber();
    Out.W = J.Items()[3].AsNumber();
    return true;
}

JsonValue TrajectorySampleToJson(const TrajectorySample& S)
{
    JsonValue Obj = JsonValue::MakeObject();
    Obj.Set("t", JsonValue(S.TimeSeconds));
    Obj.Set("pos", Vec3ToJson(S.Translation));
    Obj.Set("rot", QuatToJson(S.Rotation));
    return Obj;
}

bool TrajectorySampleFromJson(const JsonValue& J, TrajectorySample& Out)
{
    if (!J.IsObject())
    {
        return false;
    }
    Out.TimeSeconds = J.GetNumber("t");
    const JsonValue* Pos = J.Find("pos");
    const JsonValue* Rot = J.Find("rot");
    return Pos && Rot && Vec3FromJson(*Pos, Out.Translation) && QuatFromJson(*Rot, Out.Rotation);
}

namespace
{

JsonValue TimeRangeToJson(const TimeRange& R)
{
    JsonValue Obj = JsonValue::MakeObject();
    Obj.Set("startFrame", JsonValue(R.StartFrame));
    Obj.Set("endFrame", JsonValue(R.EndFrame));
    Obj.Set("fps", JsonValue(R.Fps));
    return Obj;
}

bool TimeRangeFromJson(const JsonValue* J, TimeRange& Out)
{
    if (!J || !J->IsObject())
    {
        return false;
    }
    Out.StartFrame = J->GetNumber("startFrame");
    Out.EndFrame = J->GetNumber("endFrame");
    Out.Fps = J->GetNumber("fps", 30.0);
    return true;
}

JsonValue MakeEnvelope(MessageType Type, JsonValue Payload)
{
    JsonValue Envelope = JsonValue::MakeObject();
    Envelope.Set("type", JsonValue(MessageTypeToString(Type)));
    Envelope.Set("protocolVersion", JsonValue(static_cast<double>(kProtocolVersion)));
    Envelope.Set("payload", std::move(Payload));
    return Envelope;
}

// Parses the envelope, validates type + version, and returns the payload.
bool OpenEnvelope(const std::string& PayloadJson, MessageType Expected,
                  JsonValue& OutRoot, const JsonValue*& OutPayload, std::string& OutError)
{
    if (!JsonValue::Parse(PayloadJson, OutRoot, OutError))
    {
        return false;
    }
    MessageType Type = MessageType::Unknown;
    uint32_t Version = 0;
    // Re-derive from the parsed root to avoid parsing twice.
    Type = MessageTypeFromString(OutRoot.GetString("type"));
    Version = static_cast<uint32_t>(OutRoot.GetNumber("protocolVersion"));

    if (Version != kProtocolVersion)
    {
        OutError = "Protocol version mismatch: got " + std::to_string(Version)
                 + ", expected " + std::to_string(kProtocolVersion);
        return false;
    }
    if (Type != Expected)
    {
        OutError = std::string("Unexpected message type: got '") + OutRoot.GetString("type")
                 + "', expected '" + MessageTypeToString(Expected) + "'";
        return false;
    }
    OutPayload = OutRoot.Find("payload");
    if (!OutPayload || !OutPayload->IsObject())
    {
        OutError = "Missing 'payload' object";
        return false;
    }
    return true;
}

} // anonymous namespace

bool PeekEnvelope(const std::string& PayloadJson, MessageType& OutType,
                  uint32_t& OutVersion, std::string& OutError)
{
    JsonValue Root;
    if (!JsonValue::Parse(PayloadJson, Root, OutError))
    {
        return false;
    }
    OutType = MessageTypeFromString(Root.GetString("type"));
    OutVersion = static_cast<uint32_t>(Root.GetNumber("protocolVersion"));
    return true;
}

// ---------------------------------------------------------------------------
// Handshake
// ---------------------------------------------------------------------------

std::string EncodeHandshake(const HandshakeMessage& Msg)
{
    JsonValue Payload = JsonValue::MakeObject();
    Payload.Set("clientName", JsonValue(Msg.ClientName));
    Payload.Set("characterId", JsonValue(Msg.CharacterId));
    return MakeEnvelope(MessageType::Handshake, std::move(Payload)).Write();
}

bool DecodeHandshake(const std::string& PayloadJson, HandshakeMessage& Out, std::string& OutError)
{
    JsonValue Root;
    const JsonValue* Payload = nullptr;
    if (!OpenEnvelope(PayloadJson, MessageType::Handshake, Root, Payload, OutError))
    {
        return false;
    }
    Out.ClientName = Payload->GetString("clientName");
    Out.CharacterId = Payload->GetString("characterId");
    return true;
}

std::string EncodeHandshakeAck(const HandshakeAckMessage& Msg)
{
    JsonValue Clips = JsonValue::MakeArray();
    for (const std::string& Clip : Msg.KnownClips)
    {
        Clips.Add(JsonValue(Clip));
    }
    JsonValue Payload = JsonValue::MakeObject();
    Payload.Set("serverName", JsonValue(Msg.ServerName));
    Payload.Set("knownClips", std::move(Clips));
    return MakeEnvelope(MessageType::HandshakeAck, std::move(Payload)).Write();
}

bool DecodeHandshakeAck(const std::string& PayloadJson, HandshakeAckMessage& Out, std::string& OutError)
{
    JsonValue Root;
    const JsonValue* Payload = nullptr;
    if (!OpenEnvelope(PayloadJson, MessageType::HandshakeAck, Root, Payload, OutError))
    {
        return false;
    }
    Out.ServerName = Payload->GetString("serverName");
    Out.KnownClips.clear();
    if (const JsonValue* Clips = Payload->Find("knownClips"))
    {
        for (const JsonValue& Clip : Clips->Items())
        {
            if (Clip.IsString())
            {
                Out.KnownClips.push_back(Clip.AsString());
            }
        }
    }
    return true;
}

// ---------------------------------------------------------------------------
// EvaluateRequest
// ---------------------------------------------------------------------------

std::string EncodeEvaluateRequest(const EvaluateRequest& Msg)
{
    JsonValue Target = JsonValue::MakeObject();
    Target.Set("name", JsonValue(Msg.Target.Name));
    Target.Set("pos", Vec3ToJson(Msg.Target.Translation));
    Target.Set("rot", QuatToJson(Msg.Target.Rotation));

    JsonValue Samples = JsonValue::MakeArray();
    for (const TrajectorySample& S : Msg.MayaRootSamples)
    {
        Samples.Add(TrajectorySampleToJson(S));
    }

    JsonValue Payload = JsonValue::MakeObject();
    Payload.Set("requestId", JsonValue(Msg.RequestId));
    Payload.Set("characterId", JsonValue(Msg.CharacterId));
    Payload.Set("clipName", JsonValue(Msg.ClipName));
    Payload.Set("range", TimeRangeToJson(Msg.Range));
    Payload.Set("warpWindow", TimeRangeToJson(Msg.WarpWindow));
    Payload.Set("method", JsonValue(WarpMethodToString(Msg.Method)));
    Payload.Set("target", std::move(Target));
    Payload.Set("warpRotation", JsonValue(Msg.bWarpRotation));
    Payload.Set("warpTranslation", JsonValue(Msg.bWarpTranslation));
    Payload.Set("ghostIntervalFrames", JsonValue(Msg.GhostIntervalFrames));
    Payload.Set("mayaRootSamples", std::move(Samples));
    return MakeEnvelope(MessageType::EvaluateRequest, std::move(Payload)).Write();
}

bool DecodeEvaluateRequest(const std::string& PayloadJson, EvaluateRequest& Out, std::string& OutError)
{
    JsonValue Root;
    const JsonValue* Payload = nullptr;
    if (!OpenEnvelope(PayloadJson, MessageType::EvaluateRequest, Root, Payload, OutError))
    {
        return false;
    }

    Out.RequestId = Payload->GetString("requestId");
    Out.CharacterId = Payload->GetString("characterId");
    Out.ClipName = Payload->GetString("clipName");
    Out.Method = WarpMethodFromString(Payload->GetString("method"));
    Out.bWarpRotation = Payload->GetBool("warpRotation", true);
    Out.bWarpTranslation = Payload->GetBool("warpTranslation", true);
    Out.GhostIntervalFrames = Payload->GetNumber("ghostIntervalFrames", 5.0);

    if (Out.RequestId.empty())
    {
        OutError = "EvaluateRequest missing 'requestId'";
        return false;
    }
    if (Out.Method == WarpMethod::Unknown)
    {
        OutError = "EvaluateRequest has unknown warp method '" + Payload->GetString("method") + "'";
        return false;
    }
    if (!TimeRangeFromJson(Payload->Find("range"), Out.Range))
    {
        OutError = "EvaluateRequest missing 'range'";
        return false;
    }
    if (!TimeRangeFromJson(Payload->Find("warpWindow"), Out.WarpWindow))
    {
        Out.WarpWindow = Out.Range; // window defaults to the full range
    }

    const JsonValue* Target = Payload->Find("target");
    if (!Target || !Target->IsObject())
    {
        OutError = "EvaluateRequest missing 'target'";
        return false;
    }
    Out.Target.Name = Target->GetString("name");
    const JsonValue* TargetPos = Target->Find("pos");
    const JsonValue* TargetRot = Target->Find("rot");
    if (!TargetPos || !Vec3FromJson(*TargetPos, Out.Target.Translation)
        || !TargetRot || !QuatFromJson(*TargetRot, Out.Target.Rotation))
    {
        OutError = "EvaluateRequest target transform malformed";
        return false;
    }

    Out.MayaRootSamples.clear();
    if (const JsonValue* Samples = Payload->Find("mayaRootSamples"))
    {
        for (const JsonValue& Item : Samples->Items())
        {
            TrajectorySample S;
            if (!TrajectorySampleFromJson(Item, S))
            {
                OutError = "EvaluateRequest root sample malformed";
                return false;
            }
            Out.MayaRootSamples.push_back(S);
        }
    }
    return true;
}

// ---------------------------------------------------------------------------
// EvaluateProgress
// ---------------------------------------------------------------------------

std::string EncodeEvaluateProgress(const EvaluateProgress& Msg)
{
    JsonValue Payload = JsonValue::MakeObject();
    Payload.Set("requestId", JsonValue(Msg.RequestId));
    Payload.Set("progress", JsonValue(Msg.Progress));
    Payload.Set("stage", JsonValue(Msg.Stage));
    return MakeEnvelope(MessageType::EvaluateProgress, std::move(Payload)).Write();
}

bool DecodeEvaluateProgress(const std::string& PayloadJson, EvaluateProgress& Out, std::string& OutError)
{
    JsonValue Root;
    const JsonValue* Payload = nullptr;
    if (!OpenEnvelope(PayloadJson, MessageType::EvaluateProgress, Root, Payload, OutError))
    {
        return false;
    }
    Out.RequestId = Payload->GetString("requestId");
    Out.Progress = Payload->GetNumber("progress");
    Out.Stage = Payload->GetString("stage");
    return true;
}

// ---------------------------------------------------------------------------
// EvaluateResult
// ---------------------------------------------------------------------------

std::string EncodeEvaluateResult(const EvaluateResult& Msg)
{
    JsonValue Warped = JsonValue::MakeArray();
    for (const TrajectorySample& S : Msg.WarpedTrajectory)
    {
        Warped.Add(TrajectorySampleToJson(S));
    }
    JsonValue Original = JsonValue::MakeArray();
    for (const TrajectorySample& S : Msg.OriginalTrajectory)
    {
        Original.Add(TrajectorySampleToJson(S));
    }

    JsonValue Ghosts = JsonValue::MakeArray();
    for (const GhostPose& Pose : Msg.GhostPoses)
    {
        JsonValue Joints = JsonValue::MakeArray();
        for (const JointTransform& Joint : Pose.Joints)
        {
            JsonValue J = JsonValue::MakeObject();
            J.Set("name", JsonValue(Joint.JointName));
            J.Set("pos", Vec3ToJson(Joint.Translation));
            J.Set("rot", QuatToJson(Joint.Rotation));
            Joints.Add(std::move(J));
        }
        JsonValue PoseObj = JsonValue::MakeObject();
        PoseObj.Set("t", JsonValue(Pose.TimeSeconds));
        PoseObj.Set("joints", std::move(Joints));
        Ghosts.Add(std::move(PoseObj));
    }

    JsonValue WarningsArr = JsonValue::MakeArray();
    for (const std::string& Warning : Msg.Warnings)
    {
        WarningsArr.Add(JsonValue(Warning));
    }

    JsonValue Payload = JsonValue::MakeObject();
    Payload.Set("requestId", JsonValue(Msg.RequestId));
    Payload.Set("success", JsonValue(Msg.bSuccess));
    Payload.Set("errorMessage", JsonValue(Msg.ErrorMessage));
    Payload.Set("warpedTrajectory", std::move(Warped));
    Payload.Set("originalTrajectory", std::move(Original));
    Payload.Set("ghostPoses", std::move(Ghosts));
    Payload.Set("warnings", std::move(WarningsArr));
    Payload.Set("evaluationMs", JsonValue(Msg.EvaluationMs));
    return MakeEnvelope(MessageType::EvaluateResult, std::move(Payload)).Write();
}

bool DecodeEvaluateResult(const std::string& PayloadJson, EvaluateResult& Out, std::string& OutError)
{
    JsonValue Root;
    const JsonValue* Payload = nullptr;
    if (!OpenEnvelope(PayloadJson, MessageType::EvaluateResult, Root, Payload, OutError))
    {
        return false;
    }

    Out.RequestId = Payload->GetString("requestId");
    Out.bSuccess = Payload->GetBool("success");
    Out.ErrorMessage = Payload->GetString("errorMessage");
    Out.EvaluationMs = Payload->GetNumber("evaluationMs");

    auto ReadTrajectory = [&](const char* Key, std::vector<TrajectorySample>& OutSamples) -> bool
    {
        OutSamples.clear();
        const JsonValue* Arr = Payload->Find(Key);
        if (!Arr)
        {
            return true; // optional
        }
        for (const JsonValue& Item : Arr->Items())
        {
            TrajectorySample S;
            if (!TrajectorySampleFromJson(Item, S))
            {
                OutError = std::string("EvaluateResult '") + Key + "' sample malformed";
                return false;
            }
            OutSamples.push_back(S);
        }
        return true;
    };

    if (!ReadTrajectory("warpedTrajectory", Out.WarpedTrajectory)) return false;
    if (!ReadTrajectory("originalTrajectory", Out.OriginalTrajectory)) return false;

    Out.GhostPoses.clear();
    if (const JsonValue* Ghosts = Payload->Find("ghostPoses"))
    {
        for (const JsonValue& PoseJson : Ghosts->Items())
        {
            GhostPose Pose;
            Pose.TimeSeconds = PoseJson.GetNumber("t");
            if (const JsonValue* Joints = PoseJson.Find("joints"))
            {
                for (const JsonValue& JointJson : Joints->Items())
                {
                    JointTransform Joint;
                    Joint.JointName = JointJson.GetString("name");
                    const JsonValue* Pos = JointJson.Find("pos");
                    const JsonValue* Rot = JointJson.Find("rot");
                    if (!Pos || !Vec3FromJson(*Pos, Joint.Translation)
                        || !Rot || !QuatFromJson(*Rot, Joint.Rotation))
                    {
                        OutError = "EvaluateResult ghost joint malformed";
                        return false;
                    }
                    Pose.Joints.push_back(std::move(Joint));
                }
            }
            Out.GhostPoses.push_back(std::move(Pose));
        }
    }

    Out.Warnings.clear();
    if (const JsonValue* WarningsArr = Payload->Find("warnings"))
    {
        for (const JsonValue& Warning : WarningsArr->Items())
        {
            if (Warning.IsString())
            {
                Out.Warnings.push_back(Warning.AsString());
            }
        }
    }
    return true;
}

// ---------------------------------------------------------------------------
// Error
// ---------------------------------------------------------------------------

std::string EncodeError(const ErrorMessage& Msg)
{
    JsonValue Payload = JsonValue::MakeObject();
    Payload.Set("requestId", JsonValue(Msg.RequestId));
    Payload.Set("message", JsonValue(Msg.Message));
    return MakeEnvelope(MessageType::Error, std::move(Payload)).Write();
}

bool DecodeError(const std::string& PayloadJson, ErrorMessage& Out, std::string& OutError)
{
    JsonValue Root;
    const JsonValue* Payload = nullptr;
    if (!OpenEnvelope(PayloadJson, MessageType::Error, Root, Payload, OutError))
    {
        return false;
    }
    Out.RequestId = Payload->GetString("requestId");
    Out.Message = Payload->GetString("message");
    return true;
}

} // namespace WarpViz
} // namespace AnimForge
